src/libsyntax/tokenstream.rs

```diff
@@ -417,112 +462,172 @@ fn covering_span(trees: &[TokenTree]) -> Span {
+          }
+          v
+        }
+        TokenStream::Node { ref left, ref right, .. } => {
+          let mut v1 = TokenStream::to_vec(left);
+          let mut v2 = TokenStream::to_vec(right);
+          v1.append(&mut v2);
+          v1
+        }
+      }
     }
 
-    /// Convert an interned string into a one-element TokenStream.
-    pub fn from_interned_string_as_ident(s: InternedString) -> TokenStream {
-        TokenStream::from_tts(vec![TokenTree::Token(DUMMY_SP,
-                                                    Token::Ident(token::str_to_ident(&s[..])))])
+    /// Convert a TokenStream into a vector of TokenTrees (by cloning the TokenTrees).
+    /// (This operation is expensive.)
+    pub fn to_tts(&self) -> Vec<TokenTree> {
+      TokenStream::to_vec(self).into_iter().map(|x| x.clone()).collect::<Vec<TokenTree>>()
     }
-}
 
-/// TokenSlices are 'views' of `TokenStream's; they fit the same role as `str`s do for
-/// `String`s. In general, most TokenStream manipulations will be refocusing their internal
-/// contents by taking a TokenSlice and then using indexing and the provided operators.
-#[derive(PartialEq, Eq, Debug)]
-pub struct TokenSlice([TokenTree]);
-
-impl ops::Deref for TokenStream {
-    type Target = TokenSlice;
-
-    fn deref(&self) -> &TokenSlice {
-        let tts: &[TokenTree] = &*self.tts;
-        unsafe { mem::transmute(tts) }
-    }
-}
-
-impl TokenSlice {
-    /// Convert a borrowed TokenTree slice into a borrowed TokenSlice.
-    fn from_tts(tts: &[TokenTree]) -> &TokenSlice {
-        unsafe { mem::transmute(tts) }
+    /// Manually change a TokenStream's span.
+    pub fn span(&self) -> Span {
+      match *self {
+        TokenStream::Empty => DUMMY_SP,
+        TokenStream::Leaf { sp, .. }    |
+        TokenStream::SubLeaf { sp, .. } |
+        TokenStream::Node { sp, .. }    => sp,
+      }
     }
 
-    /// Indicates whether the `TokenStream` is empty.
-    pub fn is_empty(&self) -> bool {
-        self.len() == 0
+    /// Returns an iterator over a TokenSlice (as a sequence of TokenStreams).
+    pub fn iter<'a>(&self) -> Iter {
+        Iter { vs: self, idx: 0 }
     }
 
-    /// Return the `TokenSlice`'s length.
-    pub fn len(&self) -> usize {
-        self.0.len()
+    /// Splits a TokenSlice based on the provided `&TokenTree -> bool` predicate.
+    pub fn split<P>(&self, pred: P) -> Split<P>
+        where P: FnMut(&TokenTree) -> bool
+    {
+        Split {
+            vs: self,
+            pred: pred,
+            finished: false,
+            idx: 0,
+        }
     }
 
-    /// Check equality versus another TokenStream, ignoring span information.
-    pub fn eq_unspanned(&self, other: &TokenSlice) -> bool {
-        if self.len() != other.len() {
-            return false;
+    /// Produce a slice of the input TokenStream from the `from` index, inclusive, to the
+    /// `to` index, non-inclusive.
+    pub fn slice(&self, range: ops::Range<usize>) -> TokenStream {
```

Comment:
It can't because the type requirements on `Index` demand a borrowed-out value, but the slice operations are persistent data structure operations instead.